driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java

```diff
@@ -177,53 +168,54 @@ private BsonDocument asGetMoreCommandDocument() {
+                public void onResult(final AsyncConnection connection, final Throwable t) {
+                    killCursorAsynchronouslyAndReleaseConnectionAndSource(connection, localCursor);
+                }
+            });
+        }
+    }
+
     private void killCursor(final AsyncConnection connection) {
-        if (cursor != null) {
-            final ServerCursor localCursor = cursor;
-            final AsyncConnectionSource localConnectionSource = connectionSource;
-            cursor = null;
-            connectionSource = null;
-            if (connection != null) {
-                connection.retain();
-                killCursorAsynchronouslyAndReleaseConnectionAndSource(connection, localCursor, localConnectionSource);
-            } else {
-                localConnectionSource.getConnection(new SingleResultCallback<AsyncConnection>() {
-                    @Override
-                    public void onResult(final AsyncConnection connection, final Throwable connectionException) {
-                        if (connectionException == null) {
-                            killCursorAsynchronouslyAndReleaseConnectionAndSource(connection, localCursor, localConnectionSource);
-                        }
-                    }
-                });
-            }
-        } else if (connectionSource != null) {
+        ServerCursor localCursor = cursor.getAndSet(null);
+        if (localCursor != null) {
+            killCursorAsynchronouslyAndReleaseConnectionAndSource(connection, localCursor);
+        } else {
+            connection.release();
             connectionSource.release();
-            connectionSource = null;
         }
     }
 
-    private void killCursorAsynchronouslyAndReleaseConnectionAndSource(final AsyncConnection connection, final ServerCursor localCursor,
-                                                                       final AsyncConnectionSource localConnectionSource) {
+    private void killCursorAsynchronouslyAndReleaseConnectionAndSource(final AsyncConnection connection, final ServerCursor localCursor) {
         connection.killCursorAsync(namespace, singletonList(localCursor.getId()), new SingleResultCallback<Void>() {
             @Override
             public void onResult(final Void result, final Throwable t) {
                 connection.release();
-                localConnectionSource.release();
+                connectionSource.release();
             }
         });
     }
 
     private void handleGetMoreQueryResult(final AsyncConnection connection, final SingleResultCallback<List<T>> callback,
                                           final QueryResult<T> result) {
-        if (result.getResults().isEmpty() && result.getCursor() != null) {
-            getMore(connection, callback);
+        cursor.getAndSet(result.getCursor());
+        if (isClosed() && result.getCursor() != null) {
+            killCursorAsynchronouslyAndReleaseConnectionAndSource(connection, result.getCursor());
```

Comment:
Because the cursor now is not null but close may have been called. This is the case you mentioned in the previous review that needed fixing.